// stability-ai/dist/v2beta/stable-image/control.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.style = exports.structure = exports.sketch = void 0;
const axios_1 = __importDefault(require("axios"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const form_data_1 = __importDefault(require("form-data"));
const util_1 = require("../../util");
const error_1 = require("../../error");
const Util = __importStar(require("../../util"));
const RESOURCE = 'stable-image/control';
var Endpoint;
(function (Endpoint) {
    Endpoint["SKETCH"] = "sketch";
    Endpoint["STRUCTURE"] = "structure";
    Endpoint["STYLE"] = "style";
})(Endpoint || (Endpoint = {}));
/**
 * Stability AI Stable Image Control Sketch (v2beta)
 *
 * @param image - Local filepath or public URL of the image to control sketch
 * @param prompt - Prompt to use for control sketch
 * @param options - Control sketch Options
 */
async function sketch(...args) {
    return control.call(this, Endpoint.SKETCH, ...args);
}
exports.sketch = sketch;
/**
 * Stability AI Stable Image Control Structure (v2beta)
 *
 * @param image - Local filepath or public URL of the image to control structure
 * @param prompt - Prompt to use for control structure
 * @param options - Control structure Options
 */
async function structure(...args) {
    return control.call(this, Endpoint.STRUCTURE, ...args);
}
exports.structure = structure;
async function control(endpoint, ...args) {
    const [image, prompt, options] = args;
    const imagePath = new Util.ImagePath(image);
    const formData = {
        image: fs_extra_1.default.createReadStream(await imagePath.filepath()),
        prompt,
    };
    if (options === null || options === void 0 ? void 0 : options.controlStrength)
        formData.control_strength = options.controlStrength;
    if (options === null || options === void 0 ? void 0 : options.negativePrompt)
        formData.negative_prompt = options.negativePrompt;
    if (options === null || options === void 0 ? void 0 : options.seed)
        formData.seed = options.seed;
    if (options === null || options === void 0 ? void 0 : options.outputFormat)
        formData.output_format = options.outputFormat;
    const response = await axios_1.default.postForm(Util.makeUrl(util_1.APIVersion.V2_BETA, RESOURCE, endpoint), axios_1.default.toFormData(formData, new form_data_1.default()), {
        validateStatus: undefined,
        headers: {
            ...this.authHeaders,
            Accept: 'application/json',
        },
    });
    imagePath.cleanup();
    if (response.status === 200) {
        return Util.processContentResponse(response.data, (options === null || options === void 0 ? void 0 : options.outputFormat) || Util.DEFAULT_OUTPUT_FORMAT, `v2beta_stable_image_control_${endpoint}`);
    }
    throw new error_1.StabilityAIError(response.status, `Failed to run stable image control ${endpoint}`, response.data);
}
/**
 * Stability AI Stable Image Control Style (v2beta)
 *
 * @param image - Local filepath or public URL of the image to control style
 * @param prompt - Prompt to use for control style
 * @param options - Control style Options
 */
async function style(...args) {
    const [image, prompt, options] = args;
    const imagePath = new Util.ImagePath(image);
    const formData = {
        image: fs_extra_1.default.createReadStream(await imagePath.filepath()),
        prompt,
    };
    if (options === null || options === void 0 ? void 0 : options.negativePrompt)
        formData.negative_prompt = options.negativePrompt;
    if (options === null || options === void 0 ? void 0 : options.aspectRatio)
        formData.aspect_ratio = options.aspectRatio;
    if (options === null || options === void 0 ? void 0 : options.fidelity)
        formData.fidelity = options.fidelity;
    if (options === null || options === void 0 ? void 0 : options.seed)
        formData.seed = options.seed;
    if (options === null || options === void 0 ? void 0 : options.outputFormat)
        formData.output_format = options.outputFormat;
    const response = await axios_1.default.postForm(Util.makeUrl(util_1.APIVersion.V2_BETA, RESOURCE, Endpoint.STYLE), axios_1.default.toFormData(formData, new form_data_1.default()), {
        validateStatus: undefined,
        headers: {
            ...this.authHeaders,
            Accept: 'application/json',
        },
    });
    imagePath.cleanup();
    if (response.status === 200) {
        return Util.processContentResponse(response.data, (options === null || options === void 0 ? void 0 : options.outputFormat) || Util.DEFAULT_OUTPUT_FORMAT, `v2beta_stable_image_control_style`);
    }
    throw new error_1.StabilityAIError(response.status, `Failed to run stable image control style`, response.data);
}
exports.style = style;
